k8s/src/test/scala/io/buoyant/k8s/v1/ApiTest.scala

```diff
@@ -488,13 +491,136 @@ class ApiTest extends FunSuite
+            rsp.version = req.version
+            Future.value(rsp)
+          } catch {
+            case up: TestFailedException => throw up
+            case e: Throwable =>
+              failure = e
+              Future.exception(e)
+          }
+        case _ => Future.never
+      }
+    }
+    val api = Api(service)
+
+    val (events, closable) = watch(api.withNamespace("srv").endpoints, resourceVersion = Some("1234567"))
+    val w = rsp.writer
+
+    assert(events.size == 0)
+
+    // initial event
+    await(w.write(modified0))
+    withClue("after writing one event") {
+      assert(events.size == 1)
+    }
+
+    // write an event with the same resource version: no update.
+    await(w.write(modifiedDup))
+    withClue("after writing two events with the same resource version") {
+      assert(events.size == 1)
+    }
+    closable.close()
+    if (failure != null) throw failure
+  }
+
+  test("watch allows DELETE events with the current resource version") {
+    val rsp = Response()
+    @volatile var reqCount = 0
+    @volatile var failure: Throwable = null
+    val service = FService.mk[Request, Response] { req =>
+      reqCount += 1
+      reqCount match {
+        case 1 =>
+          try {
+            assert(req.path == "/api/v1/watch/namespaces/srv/endpoints")
+            assert(req.params.getBoolean("watch").isEmpty)
+            rsp.version = req.version
+            Future.value(rsp)
+          } catch {
+            case up: TestFailedException => throw up
+            case e: Throwable =>
+              failure = e
+              Future.exception(e)
+          }
+        case _ => Future.never
+      }
+    }
+    val api = Api(service)
+    val (events, closable) = watch(api.withNamespace("srv").endpoints, resourceVersion = Some("1234567"))
+    val w = rsp.writer
+
+    assert(events.size == 0)
+    val resourceVersion1 = 4602708
+    // initial event
+    await(w.write(modified0))
+    withClue("after writing first MODIFIED event") {
+      assert(events.size == 1)
+
+      val Activity.Ok(event: EndpointsWatch) = await(events.poll())
+      assert(event.versionNum.value == resourceVersion1)
+      val EndpointsModified(eps) = event
+      assert(eps.subsets.get.flatMap(_.addresses).flatten.map(_.ip) ==
+        Seq("10.248.2.8", "10.248.7.10", "10.248.8.8"))
+    }
+
+    assert(events.size == 0)
+
+
+    // write a DELETE event with the same resource version
+    await(w.write(deleted0))
+    withClue("after writing first DELETE event") {
+      assert(events.size == 1)
+
+      val Activity.Ok(event: EndpointsWatch) = await(events.poll())
+      assert(event.versionNum.value == resourceVersion1)
+      val EndpointsDeleted(eps) = event
+      assert(eps.subsets.get.flatMap(_.addresses).flatten.map(_.ip) ==
+        Seq("10.248.2.8", "10.248.7.10", "10.248.8.8"))
+    }
+
+    // write an event with higher version than the delete event.
+    await(w.write(modified3))
```

Comment:
Shouldn't we normally see an `ADDED` after a `DELETED` before any `MODIFIED`, assuming we had an uninterrupted watch? I don't think it matters for the purposes of this test, just curious if I misunderstand how things work.

Reply:
Yeah, I believe that's true --- this should probably be an `ADDED` event. Since Linkerd handles those events in essentially the same way, I just wanted to reuse the existing test data, but it would probably be more accurate to send an `ADDED` event here.